### DISCUSSION TITLE
2nd ENET

### DISCUSSION THREAD

**Discussion by leutholl** [ssilverman/QNEthernet#80] [General]
Hi and many thanks for the wonderful library!

I'm playing around with a MicroMod Teensy and have routed the 2nd ENET pads to a LAN8720A PHY which is configured to receive the clock. This pretty much follows [that](https://forum.pjrc.com/index.php?threads/add-ethernet-to-a-micromod-format-teensy.67643/) connection. On a MicroMod only the 2nd ENET pads are exposed on the M2 connector.

How can the library be configured to use the 2nd ENET interface? Some pointers would help me.  I'm happy to share my MicroMod Teensy adaptions and hardware hookup here in return.

**Comment by ssilverman** [comment]
You will have to write a driver for it. There’s some instructions in _lwip_driver.h_. You can also follow the other two driver implementations as examples. I too wouldn’t mind seeing MicroMod support.

Basically, it all boils down to implementing all the functions from that header.

**Comment by leutholl** [comment]
I was afraid it’s that big. I was hoping that only some parts of the original T41 driver must be modified and was hoping for some hints where that would be. Should the pad redefinition be enough?!

I think nobody has done it so far….

OK let the journey begin.  Thanks

**Comment by ssilverman** [comment]
It’s a different PHY chip, and probably has a different usage. (I haven’t read docs for that one.) If you’d like to send me some hardware, I could maybe have a look at what’s required, depending on my availability.

So far, I’ve been working on this library because I’m using it and needing it myself, which partly drives how I develop for it.

**Comment by ssilverman** [comment]
It’s also very possible that the implementation won’t be too hard to adapt from the Teensy 4.1 driver, since there’s the standard RMII interface.

I think I was seeing it from the point of view of interfacing with the Teensy 4.1 MAC itself, and contrasting it to my SPI-based W5500 driver. When I think about it more, in fact, I’d be surprised if it was very difficult. I’ve just learned, over my career, “never to assume easy.” :)

**Comment by leutholl** [comment]
I totally agree. the PHY should be compatible straight forward, the registers are de-facto standardized it seems and RMII is RMII with the exception of the clock but the PHY is teensy friendly configured to receive the clock.

So it should boil down to change the pads and maybe the MUX in lwip_driver.h/cpp to use ENET2. That's where I might need help to understand what code can stay and what not. I will give that a try later today. I live in Europe and assume you're US based, not sure if sending HW is meaningful as it should be software only by now. I made a carrier board with PoE Ethernet and all sorts of peripherals. The nice thing of the MicroMod is that it exposes FlexIO 2 pins + ENET2 + Can + 2x I2C + UARTs etc.... this together with a carrier board makes it pretty cool. The board is 22mm x 22mm - and it's really small.

will let you know when I have some progress.

**Comment by ssilverman** [comment]
My thoughts on developing for hardware, eg. drivers: I disagree that it’s “software only”. It’s hard to develop software for hardware without the hardware.

**Comment by leutholl** [comment]
that's for sure. I also need the hardware.
After investing well over 12 hours I found a routing problem on my pcb which destroys the MDC and MDIO signals - I misunderstood a  reference table from the internet that I took while routing the pcb but now I see what the IOMUXC_SW_MUX_CTL_PAD with ALT8 can connect to. So I need to create my next revision of my PCB and should be must closer then.

I guess I have a basic understanding of how to port FNET to be using ENET2. It's not so straight forward as I thought but it should be possible and my hardware issue didn't help studying the code. 

I need to pause this thread until I have a new revision of the pcb or I can hack something on my faulty pcb which is what I will try tomorrow, some jumper wires might give me luck and might continue the software side.

**Reply by ssilverman** [reply]
The _QNEthernet_ library uses lwIP under the hood. I’ve already done the hard porting and integration work, and if one implements all the lwip_driver.h functions (these are my definitions), the library will just work. (That is, using an Arduino-style API.)

**Reply by leutholl** [reply]
yes this is what I will do right now. Using FNET and my two jumper wires to mitigate the pcb layout mistake, I can get link up/down.
The base address of the register for ENET2 has an offset. So this also needs to be done.
I'm confident that I can reach the goal now without waiting on a new pcb.

**Comment by leutholl** [comment]
success! I managed to get ENET2 working on MicroMod with a LAN8720A PHY. I will also design a custom board with a DP83825IRMQR PHY (the same as on the Teensy 4.1) which needs less external components (the mdi is voltage driven and it doesn't need external resistors) and it also needs other register and strap config.
For the LAN8720A an additional RESET line is needed that the main program should set towards QNEthernet. For the DP83825 two additional lines are needed. This is because on the teensy 4.1 onboard PHY these lines are predefines whereas on MircoMod the user can pick any two GPIO pins that are available on his project. The pins used on the 4.1 board are not brought out on the MicroMod so a value of the pins must be set and we can't assume defaults. For the LAN8720 with 1 RESET line I was tempted to use driver_set_chip_select_pin() which is available on the driver interface but maybe we should support Varargs on the init() to take all the pins depending on the PHY. This is a minor work put best in the hands of you as it changes things outside the driver folder.
Also I created a #define QNETHERNET_INTERNAL_DRIVER_TEENSYMM guard on my additional driver file driver_teensymm.c and driver_teensymm.h.
Then in lwip_driver.h the driver selection was extended to:
// Select a driver
#if defined(QNETHERNET_DRIVER_W5500)
#include "drivers/driver_w5500.h"
#define QNETHERNET_INTERNAL_DRIVER_W5500
#elif defined(ARDUINO_TEENSY41)
#include "drivers/driver_teensy41.h"
#define QNETHERNET_INTERNAL_DRIVER_TEENSY41
#elif defined(ARDUINO_TEENSY_MICROMOD)
#include "drivers/driver_teensymm.h"
#define QNETHERNET_INTERNAL_DRIVER_TEENSYMM
#else
#include "drivers/driver_unsupported.h"
#define QNETHERNET_INTERNAL_DRIVER_UNSUPPORTED
#endif  // Driver selection

When I have a clear guideance on how to manage the additional user pins, I can push my code.

**Reply by ssilverman** [reply]
What is the function of the two additional lines you mention?

**Reply by leutholl** [reply]
The user defined Reset Pin and on the TI PHY additionally a POWER UP Pin that must be toggled by the driver at the right moment. Both lines exist for the T41 and are hardcoded due to the PHY being onboard and predefined Pads. But on MicroMod the library must respect user defined GPIO Pins for these two funtions. 
I‘m in the process to design my other board with the TI PHY, will then adapt my driver for the TI PHY and would expect user-defined GPIOs towards the library so the init() can toggle the lines correctly (timing matters here, so it can’t be in the main program)

**Reply by ssilverman** [reply]
The `driver_set_chip_select_pin()` function is there to support the Arduino-defined `Ethernet.init(chip-select-pin)` function. In my opinion, ideally, specific pin selections should be done at the driver level, but is a bit of a grey area as to where it should actually go, and sometimes there isn't a completely "correct" answer. I would probably start with something similar to what I do in the W5500 driver, where configuration is done in a file (see _driver_w5500_config.h_). It could also be done using defines or something.

**Reply by ssilverman** [reply]
My thinking is that the hardware is unlikely to change for a given project, and so it's okay to hard-code some of these values in some configuration file or header. I also think it's unlikely that runtime control is needed for this type of configuration.

**Reply by leutholl** [reply]
ok will do so

**Comment by leutholl** [comment]
Success #2.
I managed to get the dp83825i PHY working as well. This is the same PHY as on the T4.1. This PHY has a small footprint and is harder to solder down than the LAN8720 I used previous. But the PHY is better documented and it needs only a few external components which saves board space and keeps diff pairs better impedance controlled (it has a voltage source as a MDI driver). Two new pins need to be declared to reset this PHY correctly. So this points to the same question of above but I guess some project-level defines should do the job.
As soon as I have some time, I shall consolidate all my work and make it available.

**Reply by joseangeljimenez** [reply]
Thanks @leutholl for sharing your journey on this project!
I am also working on my own custom made board for the LAN8720.
Is it possible that you share your driver code? It'll help me a lot with the firmware part of the project.
I am also willing to help you with anything you may need, or with pushing the code to the library, in case you want to.
Regards,
Jose Jimenez

**Reply by leutholl** [reply]
I'm happy to contribute my code which is working well (see below which PHY I think is best). I don't know too much on pull requests so I'd be happy that you to take my code and  make it upstream available. You need to clean the code a bit I think. in init_phy() I printed some logs and I commented the check for PHY section.
well anyway, it works very well on my board. I can't share the whole board but would be available to send some screenshots on the sensitive section. I'm using PoE frontend with extensive ESD protection for lightning and ESD discharge.

I'm now working on a PHY Link LED dimming solution as my product needs to be "readable" in sunshine and also  in very dark places. It seems that the drive current of the LED changed to a much higher current when the PHY has been reset so the brightness is no easy to control.

Over to the code:

in lwip_driver I added some switches:
`// Select a driver
#if defined(QNETHERNET_DRIVER_W5500)
#include "drivers/driver_w5500.h"
#define QNETHERNET_INTERNAL_DRIVER_W5500
#elif defined(ARDUINO_TEENSY41)
#include "drivers/driver_teensy41.h"
#define QNETHERNET_INTERNAL_DRIVER_TEENSY41
#elif defined(ARDUINO_TEENSY_MICROMOD)
  #if defined(LAN8720)
  #include "drivers/driver_teensymm8720.h"
  #define QNETHERNET_INTERNAL_DRIVER_TEENSYMM
  #elif defined(DP83825_PHY)
  #include "drivers/driver_teensymm.h"
  #define QNETHERNET_INTERNAL_DRIVER_TEENSYMM
  #endif
#else
#include "drivers/driver_unsupported.h"
#define QNETHERNET_INTERNAL_DRIVER_UNSUPPORTED
#endif  // Driver selection`

In this zip the files need to be copied to /src/drivers. The switch above will use the correct driver pair files.

the teensymm8720 pair is experimental for Microchip 8720 PHY. I think it worked (I once saw an IP ;-)), but I'm not sure. I changed to the TI DP83825 PHY as I subjected MDI integrity was a problem on my board using the Microchip (I remember link speed and auto-neg was sometimes flapping and things got a bit unstable, but the driver should be fine, probably my board or MDI termination or straps too weak etc...)

the teensymm pair is to be used for TI DP83825 PHY - which is the one I'm using now (and the same on T41) and is easier to layout and needs fewer external components. I did also strap some I/Os just to make sure it starts with autoneg etc... consult the datasheet and notice the parallel resistors to the LED pins which is different to T41). Despite the small pitch using solder paste with stencil works fine when DIYing (that was the reason to first go with the 8720 but it turned out it's not an issue to work with the TI PHY)

[driver_teensymm.zip](https://github.com/user-attachments/files/18048534/driver_teensymm.zip)

With that, I hope somebody else can use my work. We now have stable Ethernet on a Mircomod made available!

**Reply by joseangeljimenez** [reply]
Hello @leutholl and thanks a lot for sharing your driver and valuable lessons learnt during your journey.
Congratulations! You have not only managed to make the MicroMod work with one PHY but with both the Texas and Microchip variants! Double wow!

You are making me think twice about the approach for my own project. I had mostly settled on using the Microchip LAN8720A PHY, simply because I can configure the startup options with a few of pull-up/down resistors avoiding the need for using the management serial interface... However, I don't want to bump against odd issues. On the one hand, the 8720 PHY seems to be hugely popular but I have not really looked for gotchas in the Web. On the other hand, the Texas PHY is the T4.1 standard.

I have still a few weeks until I receive some protoboards for the Microchip PHY. I guess I will try both variants (Texas and Microchip), before settling on one of them. Anyhow, I will clean a little bit and test your code thoroughly, then talk to Shawn (@ssilverman) and prepare a pull request for adding support for the Teensy to the library. Surely others will happily benefit from your nice work!

**Comment by leutholl** [comment]
Cheers mate!

I recommend the TI. It’s better documented and I can help you with signal
integrity scopes if things get strange.

You can also configure the TI with straps. In fact it would be possible to
omit Management Bus completely because of the straps. You wouldn't be able to
read link up/down but it would work if you set QNEthernet link's administratively up.

The MDI is easier on the TI which helps on layouting the diff pairs. These
need additional resistors on the 8720 which could be difficult to layout.

The TI needs a reset line and one other GPIO and this is hardcoded in my
driver. That’s not so good. One should expose the #Defines to a more user
visible place.

I never had a single issue with the TI. But the 8720 gave me sleepless
nights….

Many thanks to convert this in a pull request.

José Ángel Jiménez ***@***.***> schrieb am Mo. 9. Dez. 2024
um 21:35:

**Comment by ssilverman** [comment]
There’s now a TeensyMM driver for the LAN8720A in the “teensymm” branch. It should work out of the box.

**Reply by leutholl** [reply]
Wonderful news. I'm so glad I could contribute for the TeensyMM support!
Will test your integration soon

**Comment by leutholl** [comment]
I tried the MM branch today and then realized it's not for the DP83825IRMQR PHY that I have (and 4.1 uses). My contribution above was a working driver for DP83825IRMQR. Some functions (esp. mac address getters/setter) changed from 0.28 (what I used) to 0.33 and if I try to add my hardware related changes into 0.33 I get link (L2) but never L3 working - probably because of the mac address thing. I was hoping that this branch support the DP83825IRMQR PHY and we could still work on that. We would then go back to my above question on the two additional gpio pins (POWER, RESET) and stay very close with my I/O proposal on both hardware and code.

I stumbled across that because for whatever reason my code crashed when Serial.printing in the link/addressChanged callbacks, but interestingly not there, but some lines of code later. It's as if my stack overflows when printing a bit more text this happens, when printing less text, it works.... the same code (with some prints in the callback) worked in a recent project but now doesn't anymore and I don't know if pio has updated as this is very strange. That was the reason to begin from scratch with the upstream version, but I stumbled across the "wrong" PHY.

BTW: very rarely I don't get link, after a reboot it works. hardware layout seems fine thou. I remember that this was a problem some years ago and don't know the current state. I can recognize that state in code and I want to come up with a PHY reset to mitigate the problem. Apart from that, performance is fine.

**Comment by ssilverman** [comment]
Are you saying that you want the DP83825 driver, but on ENET2?

**Reply by ssilverman** [reply]
Maybe I should explore making ENET1/2 a configurable option.

**Reply by leutholl** [reply]
yes the DP83825 PHY on ENET2 on the micromod is a popular setup and I can provide schematics on other inputs for users to recreate this combo. If you agree, then we need a define switch for the teensymm driver to initialize the DP83825.
- it needs two additional GPIO - we discussed that earlier and somewhat agreed that this can be a #define pin approach rather than an input parameter for the init/begin function as this is quite static for a given project
- I can provide code that works well (see above) - but you need to tidy it up a bit (there are printf statements reflecting dev work on my side etc...)
- I managed to use your 0.33 micromod branch with my init code but all I got is Link but never L3 comms. I suspect the change from my 0.28 state to the now current 0.33 changed something in regards to the mac address that is burnt-in in the teensy that I would ask you to have a look at it. This would explain that L3 doesn't work but link works. After that, things should work transparently.
- My rare no-link issue is something I need to chase longer also on hardware side. I suspect that the VADD rail with the ferrite bead filter can sometimes oscillate with low-ESR bypass caps when using 10nF in parallel to the higher caps. I have a hardware revision where the link was more reliable and I changed that filter - maybe for the worse - however I improved other parts other than the pcb layout - that's on my side to clear my mind. It could also be that my hardware is just fine and this no-link might be related to a topic that once popped up some years ago also on the T4.1 (BTW using the same PHY) and I lost track of what we said it could be - I remember that the reset duration has something to do with it - maybe you remember, or maybe this is unrelated to my issue.
- ENET1 on mircomod is not possible as the pins are not carried out to the m-key socket pins - micromod with ethernet MUST use ENET2 - that's why I started my journey in the first place. BTW: I also used the other PHY (the one you used) and it somewhat worked but not realiable and this I think was related to the current vs voltage driver on the MDI. The DP83825IRMQR is easier on the front-end design, requires less parts (no pull up resistors and impedance issue as the layout of these resistors are not so easy - at least not with self-solderable SMD footprints)

As for the sudden crash I found that my program is that big that I run into stack overflows and it manifested only when doing Serial.prints in the qn::Ethernet callbacks and it would crash some time later. reducing my project proved that my code was right when stack doesn't overflow. I also realized that in 0.33 the driver functions are in FLASHMEM which helps in such situation, something I didn't do for the 0.28 - I'm always a bit careful doing that with tight I/O and ISR stuff. But you seem to have no problem with that - so I follow that approach.

**Reply by leutholl** [reply]
@ssilverman any thoughts on how to continue here? I'd like to see my work on the TI PHY driver as a #define option for your next release. Also showing schematics for a teensyMM so everybody has a jump-start for their projects.

**Reply by ssilverman** [reply]
I haven’t been able to get to it yet. Sorry. :(

**Reply by ssilverman** [reply]
@leutholl I have some options for you:
1. Send me some hardware, which I would keep. I can have a look and work on the driver, but no guarantees.
2. I don't mind working on the hardware remotely, but that would be a consulting project; contact me privately for this option.

**Comment by leutholl** [comment]
I managed to patch my driver to work with post v0.28 on my micromod (custom) board - again, it uses ENET2 and the RESET and other auxiliary pins must be define by macros (this alters to any other board of course). add the file and the header (copy from t41) to the driver folder, and select the mm driver by patching driver_select.h

`// Don't include anything else if an external driver has been included
#ifndef QNETHERNET_INTERNAL_DRIVER_EXTERNAL

#if defined(QNETHERNET_DRIVER_W5500)
#include "qnethernet/drivers/driver_w5500.h"
#define QNETHERNET_INTERNAL_DRIVER_W5500
#elif defined(ARDUINO_TEENSY41)
#include "qnethernet/drivers/driver_teensy41.h"
#define QNETHERNET_INTERNAL_DRIVER_TEENSY41
#elif defined(ARDUINO_TEENSY_MICROMOD)
#include "qnethernet/drivers/driver_teensymm.h"
#define QNETHERNET_INTERNAL_DRIVER_TEENSYMM_DP83825_PHY
#else
#include "qnethernet/drivers/driver_unsupported.h"
#define QNETHERNET_INTERNAL_DRIVER_UNSUPPORTED
#endif  // Driver selection

#endif  // QNETHERNET_INTERNAL_DRIVER_EXTERNAL`

here is the driver for the DP83825 PHY.
 [
[driver_teensymm.c](https://github.com/user-attachments/files/24278250/driver_teensymm.c)
](url)
